[JIT] Add may_alias in FunctionSchema with associated tests

ttanpcs opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #81007
* #81002
* #81000
* #80984
* #80972
* __->__ #80918
* #80734

- Created may_alias method in FunctionSchema to publicize aliasing information about inputs and outputs of a schema. 
- Tested may_alias methods for basic functionality, exceptions, and wildcard functionality.

**Cases where elements of a container alias another argument will be handled with a new may_contain_alias method which will be created in a later pr**

Commit:
[JIT] Add aliasing checks in FunctionSchema with associated tests

[ghstack-poisoned]

Commit:
Update on "[JIT] Add aliasing checks in FunctionSchema with associated tests"




[ghstack-poisoned]

Commit:
Update on "[JIT] Add aliasing checks in FunctionSchema with associated tests"




[ghstack-poisoned]

Commit:
Update on "[JIT] Add aliasing checks in FunctionSchema with associated tests"


- Created AreAliasing methods in FunctionSchema to publicize aliasing information about inputs and outputs of a schema. This includes an overloaded version of AreAliasing that also adds wildcard checks into the aliasing check.
- Tested AreAliasing methods for basic functionality, exceptions, and wildcard functionality.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add aliasing checks in FunctionSchema with associated tests"


- Created AreAliasing methods in FunctionSchema to publicize aliasing information about inputs and outputs of a schema. This includes an overloaded version of AreAliasing that also adds wildcard checks into the aliasing check.
- Tested AreAliasing methods for basic functionality, exceptions, and wildcard functionality.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add aliasing checks in FunctionSchema with associated tests"


- Created AreAliasing methods in FunctionSchema to publicize aliasing information about inputs and outputs of a schema. This includes an overloaded version of AreAliasing that also adds wildcard checks into the aliasing check.
- Tested AreAliasing methods for basic functionality, exceptions, and wildcard functionality.

[ghstack-poisoned]

Gamrix commented on aten/src/ATen/core/function_schema.h:
```diff
@@ -327,6 +339,72 @@ struct FunctionSchema {
+        std::vector<TypePtr> mutable_types;
+        for (const TypePtr& inner :
+             type->expectRef<UnionType>().containedTypes()) {
+          if (auto maybe_inner_types = mapTypeToAliasTypeSet(inner)) {
+            mutable_types.insert(
+                mutable_types.end(),
+                (*maybe_inner_types).begin(),
+                (*maybe_inner_types).end());
+          }
+        }
+        if (mutable_types.size() == 0) {
+          return c10::nullopt;
+        }
+        return mutable_types;
+      }
+      case TypeKind::AnyType:
+        return {std::vector<TypePtr>{type}};
+      case TypeKind::OptionalType: {
+        auto inner = type->castRaw<OptionalType>()->getElementType();
+        return mapTypeToAliasTypeSet(inner);
+      }
+      case TypeKind::TupleType: {
+        std::vector<TypePtr> mutable_types;
+        for (const TypePtr& inner : type->expectRef<TupleType>().elements()) {
+          if (auto maybe_inner_types = mapTypeToAliasTypeSet(inner)) {
+            mutable_types.insert(
+                mutable_types.end(),
+                (*maybe_inner_types).begin(),
+                (*maybe_inner_types).end());
+          }
+        }
+        if (mutable_types.size() == 0) {
+          return c10::nullopt;
+        }
+        return {std::vector<TypePtr>{TupleType::create(mutable_types)}};
+      }
+      default:
+        return c10::nullopt;
+    }
+  }
```
Put the implementation functions like this in `function_schema.cpp` instead of `function_schema.h`

Gamrix commented on aten/src/ATen/core/function_schema.h:
```diff
@@ -373,6 +451,61 @@ struct FunctionSchema {
+    }
+    return false;
+  }
+  bool areAliasing(const SchemaArgument& lhs, const SchemaArgument& rhs, bool check_additional) const {
```
Please document what `check_additional` means.

Also, is there a need to expose two versions of this function as public?

davidberard98 commented on aten/src/ATen/core/function_schema.h:
```diff
@@ -373,6 +451,61 @@ struct FunctionSchema {
+    if (check_additional) {
+      const c10::Argument lhsArg = getCorrectList(lhs.type)[lhs.index];
+      const c10::Argument rhsArg = getCorrectList(rhs.type)[rhs.index];
+      if (lhsArg.alias_info() && lhsArg.alias_info()->isWildcardAfter()) {
+        c10::optional<std::vector<TypePtr>> lhsTypes = mapTypeToAliasTypeSet(lhsArg.type());
+        c10::optional<std::vector<TypePtr>> rhsTypes = mapTypeToAliasTypeSet(lhsArg.type());
+        if (lhsTypes && rhsTypes) {
+          std::unordered_set<TypePtr> flattened_types;
```
is this the correct logic? I'm imagining a schema like:

```
fn(Tensor x(a -> *)) -> (Tensor(b -> *), Tensor)
```

In this case I think x and output 0 should alias, but x and output 1 should not alias

goldenxuett replied:
From what I understand, x should alias both output 0 and output 1. This is because as noted in [alias_analysis.h](https://www.internalfb.com/code/fbsource/fbcode/caffe2/torch/csrc/jit/ir/alias_analysis.h) "we consider the wildcard set as potentially aliasing any value within the same type class" and both output 0 and 1 are in the tensor type class.

Gamrix replied:
I think that wildcard  set items can only alias other wildcard items. Did we ever get a confirmation from Elias about if this is the case?

davidberard98 replied:
@Krovatkin has confirmed that the second output should not alias

davidberard98 commented on aten/src/ATen/core/function_schema.h:
```diff
@@ -373,6 +451,61 @@ struct FunctionSchema {
+          }
+        }
+      }
+      if (rhsArg.alias_info() && rhsArg.alias_info()->isWildcardAfter()) {
```
could we put this into a lambda and then call the lambda twice with (lhsArg, rhsArg) and (rhsArg, lhsArg) to reduce code duplication?

davidberard98 commented on test/cpp/jit/test_schema_info.cpp:
```diff
@@ -21,9 +21,38 @@ TEST(FunctionSchemaIsMutableTest, InvalidArgument) {
+
+TEST(FunctionSchemaAreAliasingTest, Wildcard) {
+  c10::FunctionSchema schema =
+      torch::jit::getOperatorForLiteral(
```
fyi, I think you can use this for parsing arbitrary schemas:
https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/frontend/function_schema_parser.h#L13

This might be useful for:
1) parsing fake schemas
2) if you parse fake schemas instead of loading actual operators and getting their schemas, then this test won't break if those ops change.

Commit:
Update on "[JIT] Add aliasing checks in FunctionSchema with associated tests"


- Created AreAliasing methods in FunctionSchema to publicize aliasing information about inputs and outputs of a schema. This includes an overloaded version of AreAliasing that also adds wildcard checks into the aliasing check.
- Tested AreAliasing methods for basic functionality, exceptions, and wildcard functionality.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add may_alias in FunctionSchema with associated tests"


- Created may_alias method in FunctionSchema to publicize aliasing information about inputs and outputs of a schema. 
- Tested may_alias methods for basic functionality, exceptions, and wildcard functionality.

**Cases where elements of a container alias another argument will be handled with a new may_contain_alias method which will be created in a later pr**

[ghstack-poisoned]

davidberard98 commented on aten/src/ATen/core/function_schema.cpp:
```diff
@@ -8,4 +8,103 @@ void FunctionSchema::dump() const {
+  }
+}
+
+c10::optional<std::vector<TypePtr>> FunctionSchema::mapTypeToAliasTypeSet(const TypePtr& type) const {
```
Could you add a comment about where this came from and what it does

davidberard98 replied:
^ mirroring offline discussion, comments are in the .h file so this is fine.

davidberard98 commented on aten/src/ATen/core/function_schema.cpp:
```diff
@@ -8,4 +8,103 @@ void FunctionSchema::dump() const {
+  }
+}
+
+  bool FunctionSchema::may_alias(const SchemaArgument& lhs, const SchemaArgument& rhs) const {
```
this doesn't exactly match the correct behavior of wildcard aliasing, or the implementation of aliasdb. For example, in cases that return lists, `rhsArg.alias_info()` is null and we'll see no aliasing.

IMO that's fine because existing wildcard / aliasing handling is pretty hacky and it would be a lot of work to fix it all, but can you add a comment here documenting that this doesn't handle behavior correctly and in which cases this won't handle behavior correctly?

davidberard98 replied:
^ mirroring offline discussion, this is probably mostly correct (the behavior I was commenting about is more appropriate for may_contain_alias) but add a comment clarifying the difference between may_alias and may_contain_alias

davidberard98 reviewed (changes requested):
Mostly good, but could you add some comments?

Commit:
Update on "[JIT] Add may_alias in FunctionSchema with associated tests"


- Created may_alias method in FunctionSchema to publicize aliasing information about inputs and outputs of a schema. 
- Tested may_alias methods for basic functionality, exceptions, and wildcard functionality.

**Cases where elements of a container alias another argument will be handled with a new may_contain_alias method which will be created in a later pr**

[ghstack-poisoned]

goldenxuett commented:
@pytorchbot merge
